superset-frontend/src/features/tags/TagModal.tsx

```diff
@@ -0,0 +1,321 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import React, { ChangeEvent, useState, useEffect } from 'react';
+import rison from 'rison';
+import Modal from 'src/components/Modal';
+import AsyncSelect from 'src/components/Select/AsyncSelect';
+import { FormLabel } from 'src/components/Form';
+import { t, SupersetClient } from '@superset-ui/core';
+import { Input } from 'antd';
+import { Divider } from 'src/components';
+import Button from 'src/components/Button';
+import { Tag } from 'src/views/CRUD/types';
+import { fetchObjects } from 'src/features/tags/tags';
+
+interface TaggableResourceOption {
+  label: string;
+  value: number;
+  key: number;
+}
+
+export enum TaggableResources {
+  Chart = 'chart',
+  Dashboard = 'dashboard',
+  SavedQuery = 'query',
+}
+
+interface TagModalProps {
+  onHide: () => void;
+  refreshData: () => void;
+  addSuccessToast: (msg: string) => void;
+  addDangerToast: (msg: string) => void;
+  show: boolean;
+  editTag?: Tag | null;
+}
+
+const TagModal: React.FC<TagModalProps> = ({
+  show,
+  onHide,
+  editTag,
+  refreshData,
+  addSuccessToast,
+  addDangerToast,
+}) => {
+  const [dashboardsToTag, setDashboardsToTag] = useState<
+    TaggableResourceOption[]
+  >([]);
+  const [chartsToTag, setChartsToTag] = useState<TaggableResourceOption[]>([]);
+  const [savedQueriesToTag, setSavedQueriesToTag] = useState<
+    TaggableResourceOption[]
+  >([]);
+
+  const [tagName, setTagName] = useState<string>('');
+  const [description, setDescription] = useState<string>('');
+
+  const isEditMode = !!editTag;
+  const modalTitle = isEditMode ? 'Edit Tag' : 'Create Tag';
+
+  const clearResources = () => {
+    setDashboardsToTag([]);
+    setChartsToTag([]);
+    setSavedQueriesToTag([]);
+  };
+
+  useEffect(() => {
+    const resourceMap: { [key: string]: TaggableResourceOption[] } = {
+      [TaggableResources.Dashboard]: [],
+      [TaggableResources.Chart]: [],
+      [TaggableResources.SavedQuery]: [],
+    };
+
+    const updateResourceOptions = (object: Tag) => {
+      const resourceOptions = resourceMap[object.type];
+      if (resourceOptions) {
+        resourceOptions.push({
```

Comment:
Nit/style, but you can simplify this:
if (resourceOptions) {
  resourceOptions.push({

to just 
```suggestion
        resourceOptions?.push({
```